### PR TITLE
Improve API discovery memory cache

### PR DESCRIPTION
**Pull request by billiford**
This PR improves the Kubernetes API discovery memory cache. 

**TLDR;**
- removes the `Reset()` func on a mem cache client
- adds (in favor) the `NewClientForConfig(*httpclient.RestConfig)` func
- allows us a `1:1` relationship of Kubernetes host to memory cache which reduces overall API discoveries and memory usage for storing the memory caches
- stores cached entries as a struct that contains their content and createdAt value, more similar to a file caching structure

There are a couple problems with our current memory cache solution. We are caching configs, which store things like the auth bearer token, Kubernetes host, CA data, etc. We are caching these configs to see if anything has changed since the last time the memory cache client was requested, and if it did change we need to "reset" the memory cache we had stored. Another major problem is how we store (cache) the mem cache clients. We had to take into account the config host *and* timeout, so as not to return some mem cache client that has an unexpectedly long or short timeout for a given context (for example, applying manifests uses a 3 minute timeout, listing for the search page uses a 10 second timeout, etc).

This led to our cached mem cache clients storing redundant data, like:
```go
cachedMemCacheClients["host1|180"] = // the cached client for this host and timeout of 180
cachedMemCacheClients["host1|10"] = // the cached client for the same host and timeout of 10
cachedMemCacheClients["host1|60"] = // the cached client for the same host and timeout of 60
```
ALL of these cached mem clients are for the same host, and therefore hold the same API discovery data. 

To reduce redundancy this PR stores a single memory cache (not client!) for each host, then generates new clients backed by this cache for a given config. This allows us to store a memory cache with a `1:1` relationship of `clientHost:memCache`, instead of a `1:many` relationship. Now our memory caches looks like
```go
memCaches["host1"] = // memory cache store for this host
```
And we can generate a new client for a given config like
```go
memCache := memCaches[config.Host]

return memCache.NewClientForConfig(config)
```

### CODE REVIEW

**Comment by billiford** [comment]
I'm starting to think I shouldn't be calling it a "mem cache client" since we're always "making a copy" for a config. I think it would be clearer to simply refer to it as a memory cache with a function `NewClientForConfig` that returns an underlying client.

It would probably look something like this in implementation.
```go
if _, ok := memCaches[config.Host]; !ok {
	memCaches[config.Host] = memory.NewCache(ttl)
}

memCache := memCaches[config.Host]

return memCache.NewClientForConfig(config)
```
Which I think is much easier to read.

**Comment by billiford** [comment]
OK, I made the changes I commented on and updated the PR's description.

**Review by iteaguy — APPROVED** [review]
Although there's a lot of intimate functionality involved here, which I am not too familiar with yet, I could not find much to comment on. Perhaps you should ask one more person for a review as well.

**Review by PabloTheDeveloper — APPROVED** [review]
LGTM! I have some suggestions but they aren't pressing. More like just general thoughts and questions

**Comment by PabloTheDeveloper** [comment]
One last comment, the use of `mutex` in `Cache` may benefit from [`RwMutex`](https://www.sohamkamani.com/golang/mutex/#readerwriter-mutex---the-syncrwmutex-type)

File: internal/kubernetes/controller.go [hunk 1]
```diff
@@ -80,6 +78,11 @@ const (
 	ttl            = 10 * time.Minute
 )
 
+var (
+	mux       sync.Mutex
+	memCaches = map[string]*memory.Cache{}
```
[the remark this answers is not in the document]
Comment (billiford) [reply internal/kubernetes/controller.go:83]: Nah, no plans for that.  We can just hold these things in memory and not do any cleanup, similar to how a disk cache would operate. I wan't aware of that memory issue though, so thanks for posting that.

File: internal/kubernetes/cached/memory/cache.go [hunk 2]
```diff
@@ -145,135 +137,146 @@ func (d *memCacheClient) ServerGroups() (*metav1.APIGroupList, error) {
...
 }
 
 // Invalidate enforces that no cached data that is older than the current time
 // is used.
-func (d *memCacheClient) Invalidate() {
-	d.mutex.Lock()
-	defer d.mutex.Unlock()
+func (m *memCacheClient) Invalidate() {
+	m.c.mutex.Lock()
+	defer m.c.mutex.Unlock()
```
[the remark this answers is not in the document]
Comment (billiford) [reply internal/kubernetes/cached/memory/cache.go:225]: I was working with the cache as an embedded struct. I decided to give it its own field to make it explicit in what we're accessing and to show which element holds the mutex (it is very important for the cache instance to hold a mutex).

We don't follow strict rules around struct embedding in this project. We use embedding for our "controllers" which hold clients to access external services.

See the internal controller here: https://github.com/homedepot/go-clouddriver/blob/master/internal/controller.go#L24

And the embedded `v1` controller here: https://github.com/homedepot/go-clouddriver/blob/master/internal/api/v1/controller.go#L5

File: internal/kubernetes/cached/memory/cache.go [hunk 3]
```diff
@@ -43,42 +44,33 @@ type CachedDiscoveryClient interface {
...
 	invalidated bool
 	fresh       bool
+
+	c *Cache
+}
+
+// entry represents an in-memory cache of an API discovery resource.
+type entry struct {
+	Content   interface{}
+	CreatedAt time.Time
```
[the remark this answers is not in the document]
Comment (billiford) [reply internal/kubernetes/cached/memory/cache.go:64]: We could do that. It's more that would be processed behind a mutex though. Also yes, I think we need to return an error on `json.Marshal` where possible.

**Comment by billiford** [comment]
I like the idea of using an `RWMutex`, but when I started looking at it, the main place we'd get benefits would be in the `getCachedEntry` func. However, this func is also setting the `fresh` value.
```go
	m.fresh = m.fresh && ourEntry
```
So, since we're "writing" in this function we can't just do an `RLock`.

The [disk cache source code](https://github.com/kubernetes/client-go/blob/master/discovery/cached/disk/cached_discovery.go#L51) that this was copied from uses a regular Mutex. It is a bit smarter about locking and unlocking, but it needs to be since they're accessing files on disk and whatnot, which can take a while. Since we're just accessing stuff in memory I went with the lazy `defer` method.
